### PR TITLE
[d3d9] Brain-dead gamma ramp validation

### LINKED ISSUE #1988: GODS Lands of Infinity display incorrectly
GODS Lands of Infinity show incorrect display when using dxvk d3d9.dll
Screenshot_20210323_095018
It should be like this
image

### Software information
GODS Lands of Infinity

### System information
- GPU: GTX 1060
- Driver: nvidia 460.67
- Wine version: 6.4.0
- DXVK version: 1.8.1

### Apitrace file(s)
- [trace](https://www.mediafire.com/file/dq5yjfnono63oze/gods_apitrace.trace.xz/file)

### Log files
- d3d9.log: [d3d9.log.gz](https://github.com/doitsujin/dxvk/files/6186518/d3d9.log.gz)
- d3d11.log: none
- dxgi.log: none

### PR DESCRIPTION
**Pull request by DadSchoorse**
Closes #1988

### CODE REVIEW

File: src/d3d9/d3d9_swapchain.cpp [hunk 1]
```diff
@@ -668,6 +668,27 @@ namespace dxvk {
...
   }
 
+  static bool validateGammaRamp(const WORD ramp[256]) {
+    if (ramp[0] >= ramp[255]) {
+      Logger::err("validateGammaRamp: ramp inverted or flat");
+      return false;
+    }
+
+    for (size_t i = 1; i < 256; i++) {
+      std::cout << ramp[i] << std::endl;
```
Comment (Joshua-Ashton) [inline src/d3d9/d3d9_swapchain.cpp:678]: You left in logging

File: src/d3d9/d3d9_swapchain.cpp [hunk 2]
```diff
@@ -668,6 +668,27 @@ namespace dxvk {
...
+      if (ramp[i] < ramp[i - 1]) {
+        Logger::err("validateGammaRamp: ramp not monotonically increasing");
+        return false;
+      }
+      if (ramp[i] - ramp[i - 1] == UINT16_MAX) {
```
Comment (Joshua-Ashton) [inline src/d3d9/d3d9_swapchain.cpp:683]: Maybe pick something more sane likd UINT16_MAX / 2

**Review by Joshua-Ashton — CHANGES_REQUESTED** [review]

File: src/d3d9/d3d9_swapchain.cpp [hunk 3]
```diff
@@ -668,6 +668,26 @@ namespace dxvk {
     return D3D_OK;
   }
 
+  static bool validateGammaRamp(const WORD ramp[256]) {
```
Comment (Joshua-Ashton) [inline src/d3d9/d3d9_swapchain.cpp:671]: `const WORD (&ramp)[256]` to avoid pointer type decay

**Review by Joshua-Ashton — CHANGES_REQUESTED** [review]

File: src/d3d9/d3d9_swapchain.cpp [hunk 4]
```diff
@@ -677,6 +697,11 @@ namespace dxvk {
     if (unlikely(pRamp == nullptr))
       return;
 
+    if (unlikely(!validateGammaRamp(pRamp->red)
+              || !validateGammaRamp(pRamp->blue)
+              || !validateGammaRamp(pRamp->green)))
+      return;
```
Comment (doitsujin) [inline src/d3d9/d3d9_swapchain.cpp:703]: Is this correct? If I read the wine tests correctly, only one channel needs to satisfy the conditions.

Comment (DadSchoorse) [reply src/d3d9/d3d9_swapchain.cpp:703]: I changed it to only reject if all channels are not valid since it's enough for this stupid game.

**Pull request merged**